File: Code/python/optimizers.py
import sys
from pathlib import Path

build_path = Path(__file__).resolve().parents[1] / "build" / "bindings"
sys.path.append(str(build_path))

import tinytorch as tt

# ============================================================================
# Demonstrate the Optimizer module (Module 07) using AdamW on a Linear model.
# ============================================================================
print("=" * 50)
print(" TinyTorch Optimizers Demo (Module 07)")
print("=" * 50)

model = tt.Sequential()
model.add(tt.Linear(2, 4, bias=True))
model.add(tt.ReLU())
model.add(tt.Linear(4, 1, bias=True))
model.add(tt.Sigmoid())

params = model.parameters()
for param in params:
    param.requires_grad = True

# Create an AdamW optimizer over the model's parameters.
optimizer = tt.optimizers.AdamW(params, lr=1e-2, weight_decay=0.01)
print(f"\nOptimizer step_count before: {optimizer.step_count}")
print(f"lr={optimizer.lr}, beta1={optimizer.beta1}, beta2={optimizer.beta2}, "
      f"eps={optimizer.eps}, weight_decay={optimizer.weight_decay}")

criterion = tt.MSELoss()

X = tt.Tensor([0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 1.0, 1.0], [4, 2])
Y = tt.Tensor([0.0, 1.0, 1.0, 0.0], [4, 1])

epochs = 3
print("\nTraining (using AdamW)...")
for epoch in range(epochs):
    pred = model(X)
    loss = criterion(pred, Y)

    optimizer.zero_grad()   # clear previous gradients
    loss.backward()          # accumulate gradients
    optimizer.step()         # apply update

    print(f"  Epoch {epoch + 1}: loss = {loss[0]:.6f}")

print(f"\nOptimizer step_count after: {optimizer.step_count}")
print("\nAlso available: tt.optimizers.SGD and tt.optimizers.Adam")

File: Code/bindings/optimizers/optimizers.cpp
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

#include "optimizers/adam.h"
#include "optimizers/adamw.h"
#include "optimizers/optimizer.h"
#include "optimizers/sgd.h"

namespace py = pybind11;

void bind_optimizers(py::module_& m) {
    // Register the `tinytorch.optimizers` submodule.
    py::module_ opt = m.def_submodule("optimizers", "Optimization algorithms");

    // ================= Base Optimizer =================
    py::class_<Optimizer>(opt, "Optimizer")
        .def("step", &Optimizer::step)
        .def("zero_grad", &Optimizer::zero_grad)
        .def_property_readonly("step_count", &Optimizer::step_count);

    // ================= SGD =================
    py::class_<SGD, Optimizer>(opt, "SGD")
        .def(py::init<const Optimizer::ParamList&, float, float, float>(), py::arg("params"),
             py::arg("lr") = 0.01f, py::arg("momentum") = 0.0f, py::arg("weight_decay") = 0.0f)
        .def_property_readonly("lr", &SGD::lr)
        .def_property_readonly("momentum", &SGD::momentum)
        .def_property_readonly("weight_decay", &SGD::weight_decay);

    // ================= Adam =================
    py::class_<Adam, Optimizer>(opt, "Adam")
        .def(py::init<const Optimizer::ParamList&, float, float, float, float, float>(),
             py::arg("params"), py::arg("lr") = 0.001f, py::arg("beta1") = 0.9f,
             py::arg("beta2") = 0.999f, py::arg("eps") = 1e-8f, py::arg("weight_decay") = 0.0f)
        .def_property_readonly("lr", &Adam::lr)
        .def_property_readonly("beta1", &Adam::beta1)
        .def_property_readonly("beta2", &Adam::beta2)
        .def_property_readonly("eps", &Adam::eps)
        .def_property_readonly("weight_decay", &Adam::weight_decay);

    // ================= AdamW =================
    py::class_<AdamW, Adam>(opt, "AdamW")
        .def(py::init<const Optimizer::ParamList&, float, float, float, float, float>(),
             py::arg("params"), py::arg("lr") = 0.001f, py::arg("beta1") = 0.9f,
             py::arg("beta2") = 0.999f, py::arg("eps") = 1e-8f, py::arg("weight_decay") = 0.01f)
        .def_property_readonly("weight_decay", &AdamW::weight_decay);
}
